src/Feature/Registration/Feature_Registration.re

```diff
@@ -0,0 +1,406 @@
+  let macWidth = 108;
+};
+
+type viewState =
+  | Hidden
+  | EnteringKey
+  | Waiting
+  | KeySuccess
+  | KeyFailure
+  | RequestFailure;
+
+type model = {
+  licenseKey: option(string),
+  viewState,
+  inputModel: Component_InputText.model,
+};
+
+let initialInputModel =
+  Component_InputText.create(~placeholder="Enter your license key");
+
+let initial = licenseKey => {
+  licenseKey,
+  viewState: Hidden,
+  inputModel: initialInputModel,
+};
+
+let isActive = m => m.viewState != Hidden;
+let isRegistered = m => !(m.licenseKey == None);
+let getLicenseKey = m =>
+  switch (m.licenseKey) {
+  | Some(s) => s
+  | None => ""
+  };
+
+[@deriving show]
+type command =
+  | EnterLicenseKey;
+
+[@deriving show]
+type msg =
+  | Command(command)
+  | Response(Service_Registration.response)
+  | InputText(Component_InputText.msg)
+  | KeyPressed(string)
+  | Pasted(string);
+
+type outmsg =
+  | Nothing
+  | LicenseKeyChanged(string)
+  | Effect(Isolinear.Effect.t(msg));
+
+module Commands = {
+  open Feature_Commands.Schema;
+
+  let enterLicenseKey =
+    define(
+      ~category="Oni2",
+      ~title="Enter license key",
+      "oni.app.enterLicenseKey",
+      Command(EnterLicenseKey),
+    );
+};
+
+module Contributions = {
+  let commands = Commands.[enterLicenseKey];
+};
+
+let update = (model, msg) =>
+  switch (msg) {
+  | Command(EnterLicenseKey) => (
+      {...model, viewState: EnteringKey, inputModel: initialInputModel},
+      Nothing,
+    )
+  | Response(LicenseValid(key, true)) => (
+      {...model, licenseKey: Some(key), viewState: KeySuccess},
+      LicenseKeyChanged(key),
+    )
+  | Response(LicenseValid(_, false)) => (
+      {...model, viewState: KeyFailure},
+      Nothing,
+    )
+  | Response(RequestFailed) => (
+      {...model, viewState: RequestFailure},
+      Nothing,
+    )
+  | InputText(msg) =>
+    let (inputModel', _) = Component_InputText.update(msg, model.inputModel);
+    ({...model, inputModel: inputModel'}, Nothing);
+  | KeyPressed(key) =>
+    let model' =
+      switch (key) {
+      | "<CR>" when model.viewState == EnteringKey => {
+          ...model,
+          viewState: Waiting,
+        }
+      | "<ESC>" => {...model, viewState: Hidden}
+      | _ when model.viewState == EnteringKey =>
+        let inputModel =
+          Component_InputText.handleInput(~key, model.inputModel);
+        {...model, inputModel};
+      | _ => model
+      };
+
+    let outmsg =
+      switch (key) {
+      | "<CR>" when model.viewState == EnteringKey =>
+        let licenseKey = Component_InputText.value(model.inputModel);
+        let wrapEvent = evt => Response(evt);
+        Effect(
+          Service_Registration.Effect.checkLicenseKeyValidity(licenseKey)
+          |> Isolinear.Effect.map(wrapEvent),
+        );
+      | _ => Nothing
+      };
+
+    (model', outmsg);
+  | Pasted(text) =>
+    let inputModel = Component_InputText.paste(~text, model.inputModel);
+    ({...model, inputModel}, Nothing);
```

Comment:
I tested on Linux - and for some reason the `Paste` event wasn't working. I think some of the context keys might not be triggered - https://github.com/onivim/oni2/blob/92971459a0529f398f6f6cea1859511adacb10e8/src/Store/KeyBindingsStoreConnector.re#L71

I'd expect for this we'd have the `textInputFocus` context key active.

Reply:
Just pushed up a fix for this, to expose the context keys